src/workerd/api/sockets.c++

```diff
@@ -164,23 +164,34 @@ jsg::Ref<Socket> Socket::startTls(jsg::Lock& js, jsg::Optional<TlsOptions> tlsOp
 
   // The current socket's writable buffers need to be flushed. The socket's WritableStream is backed
   // by an AsyncIoStream which doesn't implement any buffering, so we don't need to worry about
-  // flushing. But this is something to keep in mind in case this assumption no longer holds in
-  // the future.
+  // flushing. But the JS WritableStream holds a queue so some data may still be buffered. This
+  // means we need to flush the WritableStream.
   //
   // Detach the AsyncIoStream from the Writable/Readable streams and make them unusable.
-  writable->removeSink(js);
-  readable = readable->detach(js, true);
-  closeFulfiller.resolver.resolve();
-
-  auto acceptedHostname = domain.asPtr();
-  KJ_IF_MAYBE(s, tlsOptions) {
-    KJ_IF_MAYBE(expectedHost, s->expectedServerHostname) {
-      acceptedHostname = *expectedHost;
+  auto& context = IoContext::current();
+  auto secureStreamPromise = context.awaitJs(writable->flush(js).then(js,
+      [this, domain = kj::heapString(domain), tlsOptions = kj::mv(tlsOptions),
```

Comment:
`domain = kj::cp(domain)` ??

Reply:
`kj::cp` for some reason doesn't work on a kj::String. I always try it first and end up going for heapString (Used to use `str` before someone told me that's not wise).

Here is the error I get:

```
bazel-out/k8-dbg/bin/external/capnp-cpp/src/kj/_virtual_includes/kj/kj/common.h:695:61: error: call to implicitly-deleted copy constructor of 'kj::String'
template<typename T> constexpr T cp(T& t) noexcept { return t; }
                                                            ^
external/workerd/src/workerd/api/sockets.c++:173:27: note: in instantiation of function template specialization 'kj::cp<kj::String>' requested here
      [this, domain = kj::cp(domain), tlsOptions = kj::mv(tlsOptions),
                          ^
bazel-out/k8-dbg/bin/external/capnp-cpp/src/kj/_virtual_includes/kj/kj/string.h:296:15: note: copy constructor of 'String' is implicitly deleted because field 'content' has a deleted copy constructor
  Array<char> content;
              ^
bazel-out/k8-dbg/bin/external/capnp-cpp/src/kj/_virtual_includes/kj/kj/array.h:146:20: note: 'Array' has been explicitly marked deleted here
  KJ_DISALLOW_COPY(Array);
                   ^
1 error generated.
```

Reply:
`kj::cp()` only works for copyable types, i.e. types that you could construct or assign by copy. Most of the time `kj::cp` is redundant since the compiler will copy automatically, but there's a few cases where it resolves ambiguity.

I usually use `kj::str()` to copy strings though `kj::heapString()` is fine too.

Probably we should add a `.clone()` method to `kj::String` (and `kj::StringPtr`, and arrays, and lots of other things...).

Reply:
oh doh, right. heh... forgot about that. A `clone()` method would be nice.